keps/0027-force-pod-reload-parameters.md

```diff
@@ -0,0 +1,203 @@
+		{"spec", "template", "metadata", "annotations"},
+		{"spec", "jobTemplate", "metadata", "annotations"},
+		{"spec", "jobTemplate", "spec", "template", "metadata", "annotations"},
+	}
+```
+The annotations are volatile and will/may change with every plan execution. The nested changes to template and jobTemplate
+are historically made because the behavior was modeled after Kustomize.
+
+This KEP proposes to change this: 
+- Attributes will only be applied to the top-level resource, not nested resources ( `spec/template/...`, `spec/jobTemplate/...`)
+- The PlanUIDAnnotation will be completely removed.
+
+The values of the attributes can still be discovered by following the ownerReference of created resources.
+
+### Risks and Mitigations
+
+The proposed change removes the automatic pod restart. It will be a breaking change, operator developers will need
+to adjust the operator to define a group for all used parameters and add their own attribute in the pod spec template
+that triggers the restart. This needs to be documented, and old operators will behave differently with the new KUDO
+version.
+
+## Implementation History
+
+- 2020-03-31 - Initial draft. (@aneumann)
+- 2020-04-14 - Add second Proposal
+- 2020-04-15 - Cleanup, moved first proposal to alternatives section
+
+## Alternatives
+
+### Full dependency graph for used resources and parameters
+
+A better alternative would be to have a full dependency graph of all resources to parameters. This would need to include
+ConfigMaps, Secrets, potentially other pods, etc. With this dependency graph, KUDO could determine when a pod restart
+is necessary and when not, i.e. changing the replica value of a stateful set would not trigger a pod restart, changing
+a value in a ConfigMap that is used by the pod template would trigger a restart.
+An operator would still want to configure if and when pod restarts may be required, as a pod can be aware of changing
+ConfigMaps. 
+
+Calculating the dependency graph would be a very complex undertaking, and may not even be possible in all cases. The
+implementation effort would be very high for only medium additional benefits
+
+### ForcePodRestart attribute for parameters
+
+Add an additional attribute `forcePodRestart` to parameter specifications in `params.yaml`:
+
+```yaml
+  - name: NODE_COUNT
+    description: "Number of Cassandra nodes."
+    default: "3"
+    forcePodRestart: "false"
+```
+
+The default value for this parameter would be `true` to keep backwards compatibility. The general behavior should stay
+the way it is right now, that on a change of a parameter the pods will be automatically restarted.
+
+If multiple parameters are changed in one update, the `forcePodRestart` flags of all attributes are `OR`ed: If at least one
+parameter has the `forcePodRestart` set to `true`, the pods will execute a rolling restart.
+
+This solution would be very easy to implement, but may be hard to explain and requires intimate knowledge of KUDO internals
```

Comment:
Well, if this is hard to explain then `{{ .ParamGroups.MAIN_GROUP.hash }}` and managing all the `GROUP`s might be even harder 😉